Do not rewrite subnamespaces or classes containing "Zend" or "ZF"

weierophinney opened the pull request:
This patch does the following:

- It changes the "edge-case" replacement maps (generally `[a-z]\\(Zend|ZF)\\`) to remove the trailing namespace separator. This allows them to match either subnamespaces _or_ classes.

- Adds additional rules for classes we specifically define within the framework that _should_ be replaced. These include:

  - `ZendAcl` from zend-expressive-authorization-acl
  - `ZendRbac` from zend-expressive-authorization-rbac
  - `ZendRouter` from zend-expressive-zendrouter
  - `ZendViewRenderer` from zend-expressive-zendviewrenderer

  In each case, I define:

  - the shortest possible match (class name without namespace)
  - ~the class name with the immediate parent namespace~ (Note: removed this, as it was a greedy lookup)
  - the fully qualified class name

  (The latter ~two variants have~ variant has two versions, to account for escaping the namespace separator.)

I've also added more test cases to verify that:

- Subnamespaces and user-defined classes do NOT get rewritten.
- Various known imports and definitions from our own packages get rewritten correctly.

Fixes #53

weierophinney committed:
fix: do not rewrite subnamespaces or classes containing "Zend" or "ZF"

This patch does the following:

- It changes the "edge-case" replacement maps (generally `[a-z]\\(Zend|ZF)\\`)
  to remove the trailing namespace separator. This allows them to match
  either subnamespaces _or_ classes.

- Adds additional rules for classes we specifically define within the
  framework that _should_ be replaced. These include:

  - `ZendAcl` from zend-expressive-authorization-acl
  - `ZendRbac` from zend-expressive-authorization-rbac
  - `ZendRouter` from zend-expressive-zendrouter
  - `ZendViewRenderer` from zend-expressive-zendviewrenderer

  In each case, I define:

  - the shortest possible match (class name without namespace)
  - the class name with the immediate parent namespace
  - the fully qualified class name

  (The latter two variants have two versions, to account for escaping
  the namespace separator.)

I've also added more test cases to verify that:

- Subnamespaces and user-defined classes do NOT get rewritten.
- Various known imports and definitions from our own packages get
  rewritten correctly.

Fixes #53

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

michalbundyra commented on config/replacements.php:
```diff
@@ -151,112 +151,138 @@
+    'x\\Zend' => 'x\\Zend',
+    'y\\Zend' => 'y\\Zend',
+    'z\\Zend' => 'z\\Zend',
+    'a\\ZendOAuth' => 'a\\ZendOAuth',
```
All these from `a\\ZendOAuth` to `z\\ZendXml` are redundant now, as we have just [a-z]\\Zend without trailing namespace separator

weierophinney replied:
Looks like length of key _does_ kick in here with regards to `strtr()` (based on 3v4l that @michalbundyra created): https://3v4l.org/dipMb

michalbundyra replied:
Yes, but in that case it does not matter - as we do not want to have replaced these phrases, so still I believe we can remove all others and keep just `[a-z]\\Zend`

michalbundyra commented on config/replacements.php:
```diff
@@ -151,112 +151,138 @@
+    // class ZendAcl in zend-expressive-authorization-acl
+    'ZendAcl' => 'LaminasAcl',
+    'Acl\\ZendAcl' => 'Acl\\LaminasAcl',
+    'Acl\\\\ZendAcl' => 'Acl\\\\LaminasAcl',
```
Why do we need double namespace separator?

weierophinney replied:
Probably overkill, but was doing it to mimic the namespace entries. But you're right - for this, not needed.

michalbundyra commented on config/replacements.php:
```diff
@@ -151,112 +151,138 @@
-    'z\\ZendXml\\' => 'z\\ZendXml\\',
+    // class ZendAcl in zend-expressive-authorization-acl
+    'ZendAcl' => 'LaminasAcl',
+    'Acl\\ZendAcl' => 'Acl\\LaminasAcl',
```
Do we have test case:
```php
'factories' => [
    'any-alias' => \Some\Vendor\Acl\ZendAcl::class,
],
```
?

weierophinney replied:
I'll add one.

weierophinney replied:
Fails. And that puts us in a strange area, because we really cannot control _every_ possible context.

I think we can only do:

- Bare name (e.g., `ZendAcl` becomes `LaminasAcl`)
- FCQN (e.g., `Zend\Expressive\Authorization\Acl\ZendAcl` becomes `Mezzio\Authorization\Acl\LaminasAcl`)

If we do the parent namespace as well, we start running into issues with other libraries too quickly.

I've just tried that approach (removing the version with the parent subnamespace), and all tests pass, including the new one you suggested, so I'm going to consider that a win. I'll push shortly.

michalbundyra replied:
And in that case migration will fail on something like that:

```php
use Zend\Expressive\Authorization\Acl;

Acl\ZendAcl::class;
```

it will be migrated to:

```php
use Mezzio\Authorization\Acl;

Acl\ZendAcl::class; // <!-- wrong, should be Acl\LaminasAcl::class;
```

we should be aware of it - maybe add it into FAQ of the migration?

michalbundyra replied:
Also another case:
```php
use Some\Vendor\ZendAcl;

ZendAcl::class;
```

after migration it becomes:

```php
use Some\Vendor\ZendAcl;

LaminasAcl::class; // <-- wrong as this class does not exist, it should be ZendAcl::class
```

As in the previous comment, probably we cannot cover all cases so we should only add it into FAQ or Migration notes/issues.

More important is that PostProcessor replaces classes correctly and I believe we have achieved it here.

weierophinney replied:
Agreed - these will need to be part of the FAQ.

I don't think they'll happen often, but they _can_ happen. I added test cases, and verified that they fail, and then removed them, as we expect those failures currently.

michalbundyra replied:
@weierophinney created issue for that: https://github.com/laminas/laminas.github.io/issues/63

weierophinney committed:
fix: loosen lookups for Zend* class names

Looking up the immediate parent namespace is problematic, as a user
might define using the same hierarchy. As such, this patch reduces the
edge-case ZF lookups to short name and FQCN only.

It also adds additional test cases to ensure these scenarios do not
result in unnecessary replacements.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
fix: remove redundant maps

- [a-z]\\ZendOAuth
- [a-z]\\ZendService
- [a-z]\\ZendXml

(Redundant, as [a-z]\\Zend catches these already.)

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

michalbundyra commented on config/replacements.php:
```diff
@@ -151,112 +151,52 @@
+    // class ZendAcl in zend-expressive-authorization-acl
+    'ZendAcl' => 'LaminasAcl',
+    'Zend\\Expressive\\Authorization\\Acl\\ZendAcl' => 'Mezzio\\Authorization\\Acl\\LaminasAcl',
+    'Zend\\\\Expressive\\\\Authorization\\\\Acl\\\\ZendAcl' => 'Mezzio\\\\Authorization\\\\Acl\\\\LaminasAcl',
```
Do we need this with duplicated namespace separators?

The same for couple more below.

weierophinney replied:
Yes, because they could be in code as string values, and users often escape those (same reason we do double-escaped variants of namespaces above).

michalbundyra replied:
There is another suggestion:
What if we keep just "single" namespace separator in that file and in https://github.com/laminas/laminas-zendframework-bridge/blob/master/src/Replacements.php
we add additional with double and quadruple namespace separators?

The file here will be much cleaner and it will solve also the other problem: 
https://github.com/laminas/laminas-migration/issues/26

I am fine with addressing it separately.

weierophinney replied:
I've just pushed a commit with this change.

weierophinney committed:
refactor: simplify how NS escaping occurs

Instead of having duplicate versions of strings in the configuration
file, @michalbundyra has suggested we should perform NS escaping in the
`Replacements` class instead. This (a) ensures we do not forget to add
the escaped version of the string, and (b) simplifies the configuration.

Fixes laminas/laminas-migration#26

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

michalbundyra reviewed (approved):
🚢

weierophinney committed:
docs: Adds CHANGELOG entries for #56

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
